### PR TITLE
change type of numerical architecture symbols to integer

### LINKED ISSUE #5878: [RFC] CBMC architecture configuration uses architecture-specific int width
CBMC version: 5.24 (b26d347967)
Operating system: any
Exact command line resulting in the issue: n/a
What behaviour did you expect: Linking together goto binaries built on different platforms is possible
What happened instead: When writing goto binaries, CBMC uses `int` to create constants like `__CPROVER_architecture_int_width`. It's not obvious that such platform configuration options should be in the symbol table at all, and if they should live in there, their types should likely not be architecture-specific. `configt` places no requirement on those types, and could easily deal with `ID_integer` typed constants.

This is mostly an RFC and a note for possible future clean-up. There is no currently-known use case that's blocked by the existing implementation.

### PR DESCRIPTION
**Pull request by kroening**
Following a suggestion in ##5878, this changes the type of the numerical __CPROVER_architecture_ symbols from `int` (which is itself architecture-dependent) to integer (which does not depend on the architecture).

Closes #5878

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
Sure, we could do this, but it will require updating the goto binary version (and re-compiling binaries that are part of regression tests).

**Comment by kroening** [comment]
> Sure, we could do this, but it will require updating the goto binary version (and re-compiling binaries that are part of regression tests).

No need -- this is compatible both ways (old versions can read new binaries, and old binaries can be read by new versions).

**Comment by tautschnig** [comment]
> > Sure, we could do this, but it will require updating the goto binary version (and re-compiling binaries that are part of regression tests).
> 
> No need -- this is compatible both ways (old versions can read new binaries, and old binaries can be read by new versions).

Linking won't work for the architecture symbols now have conflicting types. We might gloss over that and would then just need to recompile the goto binaries in our regression tests.

Further fixes that seem to be necessary:
1. Intervals don't know about `integer_typet`, which can be fixed as follows:
````patch
--- a/src/util/interval.cpp
+++ b/src/util/interval.cpp
@@ -1094,7 +1094,7 @@ bool constant_interval_exprt::is_numeric(

 bool constant_interval_exprt::is_int(const typet &type)
 {
-  return (is_signed(type) || is_unsigned(type));
+  return is_signed(type) || is_unsigned(type) || type.id() == ID_integer;
 }

 bool constant_interval_exprt::is_float(const typet &src)
````
2. The C++ front-end rejects the type conversion in the assignment of the architecture string.

File: src/ansi-c/ansi_c_internal_additions.cpp [hunk 1]
```diff
@@ -113,7 +113,8 @@ static std::string architecture_string(const std::string &value, const char *s)
 template <typename T>
 static std::string architecture_string(T value, const char *s)
 {
-  return std::string("const int " CPROVER_PREFIX "architecture_") +
+  return std::string("const __CPROVER_integer " CPROVER_PREFIX
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_internal_additions.cpp]: I think this should also use `CPROVER_PREFIX` in the first case of "\_\_CPROVER_" (actually the failing cpplint job says the same :-) )

**Review by tautschnig — APPROVED** [review]

**Pull request merged**